extensions/amp-analytics/0.1/amp-analytics-3p-message-queue.js

```diff
@@ -0,0 +1,188 @@
+ */
+
+import {dev} from '../../../src/log';
+import {AMP_ANALYTICS_3P_MESSAGE_TYPE} from '../../../src/3p-analytics-common';
+import {throttle} from '../../../src/utils/rate-limit';
+
+/** @private @const {string} */
+const TAG_ = 'amp-analytics.CrossDomainIframeMessageQueue';
+
+/** @private @const {number} */
+const MESSAGE_THROTTLE_TIME_ = 100;
+
+/** @private @const {number} */
+const MAX_QUEUE_SIZE_ = 100;
+
+/**
+ * @visibleForTesting
+ * @abstract
+ */
+class AbstractAmpAnalytics3pMessageQueue {
+  /**
+   * Constructor
+   * @param {!Window} win The window element
+   * @param {!../../../3p/iframe-messaging-client.IframeMessagingClient}
+   *   iframeMessagingClient Facilitates cross-frame communication
+   * @param {!string} messageType The type that will be used to
+   * contain the queued data (see 3p-analytics-common)
+   */
+  constructor(win, iframeMessagingClient, messageType) {
+    /** @type {boolean} */
+    this.isReady_ = false;
+
+    /** @type {!Window} */
+    this.win_ = win;
+
+    /** @type {!../../../3p/iframe-messaging-client.IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @type {string} */
+    this.messageType_ = messageType;
+
+    /** @type {!Object<string,!string|!Array<string>>} */
+    this.store_ = {};
+
+    this.throttledFlushQueue_ = throttle(this.win_, () => {
+      this.flushQueue_();
+    }, MESSAGE_THROTTLE_TIME_);
+
+  }
+
+  /**
+   * Indicate that a cross-domain frame is ready to receive messages, and
+   * send all messages that were previously queued for it.
+   * associated data
+   */
+  setIsReady() {
+    this.isReady_ = true;
+    this.throttledFlushQueue_();
+  }
+
+  /**
+   * Send queued data (if there is any) to a cross-domain iframe
+   * @private
+   */
+  flushQueue_() {
+    if (this.isReady_ && Object.keys(this.store_).length > 0) {
+      this.iframeMessagingClient_./*OK*/sendMessage(this.messageType_,
+        this.buildMessage_());
+      this.store_ = {};
+    }
+  }
+
+  /**
+   * @return {../../../src/3p-analytics-common.AmpAnalytics3pNewCreative|
+   *          ../../../src/3p-analytics-common.AmpAnalytics3pEvent}
+   * @abstract
+   */
+  buildMessage_() {
+  }
+}
+
+export class AmpAnalytics3pNewCreativeMessageQueue extends
+  AbstractAmpAnalytics3pMessageQueue {
+
+  /**
+   * Constructor
+   * @param {!Window} win The window element
+   * @param {!../../../3p/iframe-messaging-client.IframeMessagingClient}
+   *   iframeMessagingClient Facilitates cross-frame communication
+   */
+  constructor(win, iframeMessagingClient) {
+    super(win, iframeMessagingClient, AMP_ANALYTICS_3P_MESSAGE_TYPE.CREATIVE);
+  }
+
+  /**
+   * Enqueues a message (event or extra data) to be sent to a cross-domain
+   * iframe.
+   * @param {!string} senderId Identifies which creative is sending the message
+   * @param {string=} opt_data The data to be enqueued and then sent to the
+   * iframe
+   */
+  enqueue(senderId, opt_data) {
```

Comment:
Intentional. They implement it differently, since for new creatives each entry in .creativeToPendingMessages_ is <string, string> since you can only have one extraData per creative, whereas for events it's <string, Array<string>> since one creative can have any number of events. I originally had just the parent class with no subclasses, and this is actually what made me split it into the current class structure.